Fix all runtime IL generation paths from passing GC types to unmanaged function calls.

AaronRobinsonMSFT opened the pull request:
Fixes #34279 

**Update:** The assert added in this PR impacts more than first thought. When a P/Invoke, call in IL, is inlined any native arguments must be converted (`conv.i`/`conv.u`) prior to the call since the IL marshaler fixes in this PR won't run. See https://github.com/dotnet/runtime/issues/35187 for an example.

/cc @jkoritzinsky @elinor-fung @jkotas @AndyAyersMS

AaronRobinsonMSFT commented:
@davidwrighton and @MichalStrehovsky for crossgen2 changes.

jkotas commented on src/coreclr/src/vm/ilmarshalers.cpp:
```diff
@@ -2907,7 +2907,7 @@ MarshalerOverrideStatus ILSafeHandleMarshaler::ArgumentOverride(NDirectStubLinke
 
             // Leave the address of the native handle local as the argument to the native method.
-            pslILDispatch->EmitLDLOCA(dwNativeHandleLocal);
+            EmitLoadNativeLocalAddrForByRefDispatch(pslILDispatch, dwNativeHandleLocal);
```
Nit: I am not sure whether factoring this into a helper method helps with readability. `EmitLDLOCA` + `EmitCONV_I` inline may be better. (matter of preference)

AaronRobinsonMSFT replied:
Not sure that it does either. I started down that path because we missed this in so many places and I was frustrated with the same two lines being in 4 places. My attempt was to make the method name verbose. I am going to leave it as such for now. I am hoping we can avoid touching this again entirely with the source generator approach.

AaronRobinsonMSFT commented:
Le sigh... Dynamic runtime `IDispatch::Invoke()` support 😞

AaronRobinsonMSFT committed:
Native COM clients have not been running since we switched over to SDK projects.

AaronRobinsonMSFT commented:
... and IJW ...

Also it looks like during the switch over to SDK test projects we inadvertently disabled all native COM client tests for consuming managed COM servers. I am re-enabling them in this PR but will rebase when I merge so the two commits exist.

AaronRobinsonMSFT commented:
... and WinRT ...

AaronRobinsonMSFT committed:
Stop IL generators from passing GC types to unmanaged function calls.
Update crossgen2 IL stub generators.
Update Dynamic runtime stub generator.
Add assert for GC types in unmanaged function calls.

BruceForstall commented on src/coreclr/src/jit/importer.cpp:
```diff
@@ -6738,9 +6738,16 @@ void Compiler::impPopArgsForUnmanagedCall(GenTree* call, CORINFO_SIG_INFO* sig)
+        GenTree* arg = argUse.GetNode();
+        if (varTypeIsGC(arg->TypeGet()))
+        {
+            assert(!"*** invalid IL: gc type passed to unmanaged call");
```
Is it ok for this to be ignored in Release builds? Or should this be a BADCODE?

AaronRobinsonMSFT replied:
That is a good question. I took the guidance from https://github.com/dotnet/runtime/issues/34279#issuecomment-613695776.

jkotas replied:
Given how many places we had to fix in our stuff, I would be very worried about breaking potential of turning this into BADCODE and the long tail of issues from that. It may be better to look into making JIT resilient to this.
